lib/tlog/tap.c

```diff
@@ -206,21 +212,70 @@ tlog_tap_setup(struct tlog_errs **perrs,
         }
 
         /* Fork a child connected via a PTY */
-        tap.pid = forkpty(&master_fd, NULL, &tap.termios_orig, &winsize);
+        tap.pid = forkpty(&master_fd, tty_path, &tap.termios_orig, &winsize);
         if (tap.pid < 0) {
             grc = TLOG_GRC_ERRNO;
             TLOG_ERRS_RAISECS(grc,
                               "Failed forking a child connected via a PTY");
         }
         /* If parent */
         if (tap.pid != 0) {
+            int ofs;
+            struct timeval tv;
+            struct utmpx *res;
+            char *ses_line;
+
             /* Split master FD into input and output FDs */
             tap.in_fd = master_fd;
             tap.out_fd = dup(master_fd);
             if (tap.out_fd < 0) {
                 grc = TLOG_GRC_ERRNO;
                 TLOG_ERRS_RAISECS(grc, "Failed duplicating PTY master FD");
             }
+
+            memset(&tap.ut, 0, sizeof(tap.ut));
+
+            ses_line = ttyname(STDIN_FILENO);
+            if (ses_line) {
+                struct utmpx *ut;
+
+                setutxent();
+
+                /* Find the current session utmp entry, if any */
+                ofs = strncmp(ses_line, TAP_DEVPATH, TAP_DEVSIZE) == 0 ? TAP_DEVSIZE : 0;
+                ses_line += ofs;
+                while ((ut = getutxent()) != NULL &&
+                        (ut->ut_type != USER_PROCESS || strcmp(ut->ut_line, ses_line)))
+                    ;
+
+                tap.ut.ut_type = USER_PROCESS;
+                tap.ut.ut_pid = tap.pid;
+
+                /* Strip the leading "/dev/" to get the line name */
+                ofs = strncmp(tty_path, TAP_DEVPATH, TAP_DEVSIZE) == 0 ? TAP_DEVSIZE : 0;
+                strncpy(tap.ut.ut_line, tty_path + ofs, TAP_LINESIZE - 1);
+
+                /* Copy the trailing pty number to ut_id */
+                ofs = strlen(tty_path) - sizeof(tap.ut.ut_id);
+                if (ofs > 0) {
+                    memcpy(tap.ut.ut_id, tty_path + ofs, sizeof(tap.ut.ut_id));
+                }
+
+                /* Set the time */
+                gettimeofday(&tv, NULL);
+                tap.ut.ut_tv.tv_sec = tv.tv_sec;
+                tap.ut.ut_tv.tv_usec = tv.tv_usec;
+
+                strcpy(tap.ut.ut_user, getpwuid(getuid())->pw_name);
+                if (ut) {
+                    strcpy(tap.ut.ut_host, ut->ut_host);
+                    tap.ut.ut_session = ut->ut_session;
+                    memcpy(tap.ut.ut_addr_v6, ut->ut_addr_v6, sizeof(tap.ut.ut_addr_v6));
+                }
+
+                res = pututxline(&tap.ut);
+                endutxent();
+            }
```

Comment:
Can you please add all the code above into its own function? Something like `tlog_tap_utmp_entry_add`

Reply:
adding a static “tlog_tap_add_utmp_entry” function

Reply:
I'm also adding a corresponding “tlog_tap_remove_utmp_entry”